### PR TITLE
Add ARM native build

### PR DESCRIPTION
**Pull request by t-h-e**
This PR adds support for building x64 and arm64 explicitly.

image

The *.appxbundle should contain both *.exe. I was not able to install the bundle due to 'This app package’s publisher certificate could not be verified. ...' even though I add the certificate to my trust store.

The *.appxbundle was create within visual studio. I do not know if the azure pipeline builds correctly though.

Issue https://github.com/tom-englert/RegionToShare/issues/80

### CODE REVIEW

**Comment by t-h-e** [comment]
1. I saw that. Actually I'm not sure if it is possible to build an ARM version without ARM CPU. While .Net 4.8.1 added ARM support, I'm not sure if cross compilation works before .NET 5
2. You can use ["Batch Build"](https://learn.microsoft.com/en-us/visualstudio/ide/how-to-build-multiple-configurations-simultaneously?view=vs-2022) in visual studio to build multiple configurations. Building the app bundle with multiple configurations is defined here: https://github.com/tom-englert/RegionToShare/pull/81/files#diff-e9ec4d05fea863d562153323044476c60887dc1828eb824b42983742e08960beR45
3. What do you mean with package? An exe can only be built for one platform. To my understanding the app bundle will contain the executables for each configuration
  a. Locally built app bundle image
4. Good point. I haven't thought about this.


Note that I have not did any cross compilation or actually any packaging of .net applications so far. This is my first attempt.

**Comment by t-h-e** [comment]
@tom-englert Pipeline is green and according to the logs x64 as well as ARM64 platform have been built.

Only x64 is published in the first `PublishBuildArtifacts` task.
The app bundle published in the second `PublishBuildArtifacts` task  should contain both platforms. At least that is what I expect.
Can I somehow check the app bundle?

**Comment by t-h-e** [comment]
Found the artifacts: https://dev.azure.com/tom-englert/Open%20Source/_build/results?buildId=6450&view=artifacts&pathAsName=false&type=publishedArtifacts

I installed that version and it is indeed running ARM64 natively 🥳 
I have not tested x64 yet

**Comment by t-h-e** [comment]
I checked on x64 and it works as well. You can even open the *.appxbundle with 7-Zip
image

**Comment by t-h-e** [comment]
Regarding
> none of the dependent packages is available as Arm64, so I wonder what would be the benefit of having half of the application Arm64 specific, while the rest is still AnyCPU.

As far as I understand, AnyCPU is exactly what you want for the rest of the application as it can be used by any architecture. I assume that as dotnet compiles code to IL, the JIT compiler will take care of the compilation to the correct machine code.
Only the executeable itself is platform specific. And even there most of the binary is identical (probably because it is IL as well). Windows needs to know how to start the application and if it should be emulated as x64 or run natively on arm64 (in case of arm).

In summary, it does not matter that the rest of the application is AnyCPU. The important part is that the app runs natively for better performance.

The emulation is nice but not nearly as fast as native code.

**Comment by tom-englert** [comment]
> Windows needs to know how to start the application and if it should be emulated as x64 or run natively on arm64 (in case of arm).

I see, but is Windows really so stupid that if it finds an AnyCPU executable that it emulates x64 on arm instead of running arm64 native?

**Comment by t-h-e** [comment]
It seems I have mixed up .Net Framework and .Net (Core) when researching this topic.
.Net (Core) needs two separate executables, as it is not as deeply integrated into the OS as .Net Framework and the structure of the executable changed.
.Net Framework can still run natively with AnyCPU on any architecture, but by default AnyCPU runs as x64 on Windows on ARM. To tell Windows or dotnet runtime to run arm64 natively, one needs to add the [`<supportedArchitecture/>`](https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#supportedarchitectures) attribute.

See https://github.com/dotnet/core/issues/7709 for a lot more explanation.

Even Microsoft seems to get confused from time to time, if something is for .Net Framework or .Net (Core): https://github.com/MicrosoftDocs/win32/pull/1855/files

**Comment by tom-englert** [comment]
That's a much better solution 👍
